src/Console/CommandCollection.php

```diff
@@ -130,4 +133,62 @@ public function count()
     {
         return count($this->commands);
     }
+
+    /**
+     * Automatically discover shell commands in CakePHP, the application and all plugins.
+     *
+     * Commands will be located using filesystem conventions. Commands are
+     * discovered in the following order:
+     *
+     * - CakePHP provided commands
+     * - Application commands
+     * - Plugin commands
+     *
+     * Commands from plugins will be added based on the order plugins are loaded.
+     * Plugin shells will attempt to use a short name. If however, a plugin
+     * provides a shell that conflicts with CakePHP or the application shells,
+     * the full `plugin_name.shell` name will be used. Plugin shells are added
+     * in the order that plugins were loaded.
+     *
+     * @return $this
+     */
+    public function autoDiscover()
```

Comment:
What if autodiscover returned the list of commands instead. That would make the api look like

```php
$collection->addAll($collection->autoDiscover());
```

The benefit I see from doing this is that you could filter-out or reorder some of the results if needed.